contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -0,0 +1,329 @@
+func KObj(obj metav1.Object) ObjectRef
+
+// KRef is used to create ObjectRef when logging information about Kubernetes objects without access to metav1.Object
+// Examples:
+// >> klog.InfoS("Pod status updated", "pod", klog.KRef(podNamespace, podName), "status", "ready")
+// output:
+// >> I1025 00:15:15.525108       1 controller_utils.go:116] "Pod status updated" pod="kube-system/kubedns" status="ready"
+func KRef(namespace, name string) ObjectRef
+
+// ObjectRef represents a reference to a kubernetes object used for logging purpose
+// In text logs it is serialized into "{namespace}/{name}" or "{name}" if namespace is empty
+type ObjectRef struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace,omitempty"`
+}
+```
+
+## Migration
+
+1. Change log functions to structured equivalent
+1. Remove string formatting from log message
+1. Name arguments
+1. Use `klog.KObj` and `klog.KRef` for Kubernetes object references
+1. Compare outputted log
+
+## Change log functions to structured equivalent
+
+Structured logging functions follow different logging interface design then other functions in `klog`. They follow
+minimal design from [logr] thus there is no 1to1 mapping.
+
+
+Simplified mapping between functions:
+* `klog.Infof`, `klog.Info`, `klog.Infoln`, `klog.InfoDepth` -> `klog.InfoS`
+* `klog.V(N).Infof`, `klog.V(N).Info`, `klog.V(N).Infoln` -> `klog.V(N).InfoS`
+* `klog.Warning`, `klog.Warningf`, `klog.Warningln`, `klog.WarningDepth` -> `klog.InfoS`
+* `klog.Error`, `klog.Errorf`, `klog.Errorln`, `klog.ErrorDepth` -> `klog.ErrorS`
+* `klog.Fatal`, `klog.Fatalf`, `klog.Fatalln`, `klog.FatalDepth` -> `klog.ErrorS`
+
+### Removing Depth
+
+Functions with depth (`klog.InfoDepth`, `klog.WarningDepth`, `klog.ErrorDepth`, `klog.FatalDepth`) are used to indicate
+that source of log (added as metadata in log) is different then invocation of logging library. This is usually used when
+implementing logging util functions. In those situations util functions should be removed and `klog` call moved up.
+
+For example
+```go
+func Handle(w http.ReponseWriter, r *http.Request) {
+    logHTTPRequest(r)
+    handle(w, r)
+}
+func logHTTPRequest(r *http.Request) {
+    klog.InfoDepth(1, "Got request")
+}
+```
+should be replaced with
+```go
+func Handle(w http.ReponseWriter, r *http.Request) {
+    klog.Info("Got request")
+    handle(w, r)
+}
+```
+
+### Using ErrorS
+
+With `klog` structured logging borrowing interface from [logr] it also inherits it's differences in semantic of error
+function. This is visible in `klog.ErrorS` which requires passing an `error` as first argument. This is because in
+[logr] Error is used to inform users about errors received from subordinate function calls.
+This means that `klog.ErrorS` should only be used when there is a obvious Golang `error` object available.
+
+Creating error just to call `klog.ErrorS` or passing `nil` is discouraged, please consider using `klog.InfoS` instead.
```

Comment:
Based on the klog implementation if nil is passed as error ErrorS implementation will act as InfoS.

Reply:
Should we mention this?

Reply:
As this is an implementation detail I would not mentioned it because someone may start relying on it. Instead I would state that this is forbiden.